src/Tinfoil/Data/Hash.hs

```diff
@@ -0,0 +1,26 @@
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE DeriveGeneric #-}
+module Tinfoil.Data.Hash(
+    Hash(..)
+  , hexDigest
+) where
+
+import           Data.ByteString (ByteString)
+import qualified Data.ByteString.Base16 as Base16
+import qualified Data.Text.Encoding as T
+
+import           GHC.Generics (Generic)
+
+import           P
+
+-- | Binary representation of a hash.
+newtype Hash =
+  Hash {
+    unHash :: ByteString
+  } deriving (Eq, Show, Generic)
+
+instance NFData Hash
```

Comment:
I think it depends on how quickly we're planning to switch to 7.10 - afaict, the implementation in 7.8.4 is pretty broken, I've been adding these instances so benchmarks compile rather than because I'm actually relying on them to evaluate to NF rather than WHNF as it hasn't mattered thus far. I'd say it's worth implementing them manually for safety/clarity, will fix these up and add a PR on guideline.